### PR TITLE
[Hold for 1.107] create global security contexts for nonRoot

### PR DESCRIPTION
**Pull request by JesseGoodier**
## What does this PR change?

Change global setting to run all pods as nonRoot and adds ability to make global changes to securityContexts.

Exceptions:
1. networkCosts requires root, is optional and not enabled by default
2. queryService initContainer requires root, is optional and not enabled by default
3. NodeExporter requires hostNetwork, is optional and is enabled by default
4. KSM should be disabled

New defaults (exceptions above):
```yaml
global:
  securityContext:
    runAsNonRoot: true
    seccompProfile:
      type: RuntimeDefault
    fsGroup: 65534
    runAsGroup: 65534
    runAsUser: 65534
  containerSecurityContext:
    allowPrivilegeEscalation: false
    readOnlyRootFilesystem: true
    capabilities:
      drop:
      - ALL
```

## Does this PR rely on any other PRs?

no

## How does this PR impact users? (This is the kind of thing that goes in release notes!)

change global setting to run all pods as nonRoot and adds ability to make global changes to securityContexts:

## Links to Issues or ZD tickets this PR addresses or fixes

https://kubecost.atlassian.net/browse/CORE-274

## How was this PR tested?

test cluster with QSR and federator enabled

## Have you made an update to documentation?

will be in values, but could use an official doc.

### CODE REVIEW

**Comment by chipzoller** [comment]
Since this PR gets almost all the way there, we may want to consider adding the additional two fields necessary to fully meet the [Pod Security Standards restricted profile](https://kubernetes.io/docs/concepts/security/pod-security-standards/#restricted). These standards are increasingly important for users who require policy enforcement across the cluster, whether that be with Kubernetes' built-in [Pod Security Admission](https://kubernetes.io/docs/concepts/security/pod-security-admission/) or dynamic admission controllers. The two missing fields, both under the `containers[]` array, are `privileged` and `allowPrivilegeEscalation`. Both should be set to a value of `false` for every container in the Pod.

**Comment by jessegoodier** [comment]
> Since this PR gets almost all the way there, we may want to consider adding the additional two fields necessary to fully meet the [Pod Security Standards restricted profile](https://kubernetes.io/docs/concepts/security/pod-security-standards/#restricted). These standards are increasingly important for users who require policy enforcement across the cluster, whether that be with Kubernetes' built-in [Pod Security Admission](https://kubernetes.io/docs/concepts/security/pod-security-admission/) or dynamic admission controllers. The two missing fields, both under the `containers[]` array, are `privileged` and `allowPrivilegeEscalation`. Both should be set to a value of `false` for every container in the Pod.

Awesome!
can you add where appropriate? 

We aren't going to merge this until 107, so we have a couple weeks.

File: cost-analyzer/values.yaml [hunk 1]
```diff
@@ -217,8 +217,19 @@ global:
...
-    # readOnlyRootFilesystem: true
+  securityContext:
+    runAsNonRoot: true
+    seccompProfile:
+      type: RuntimeDefault
+    fsGroup: 65534
+    runAsGroup: 65534
+    runAsUser: 65534
+  containerSecurityContext:
+    allowPrivilegeEscalation: false
```
Comment (chipzoller) [inline cost-analyzer/values.yaml:228]: After looking at this more carefully, you've already got `allowPrivilegeEscalation`. It'd probably be easiest for you to just add a second line `privileged: false` beneath it.

**Review by chipzoller — CHANGES_REQUESTED** [review]
Rather than me pushing one line, this looks like all that's needed for the cost-analyzer values.

File: cost-analyzer/values.yaml (hunk 1 above)
Comment (jessegoodier) [reply cost-analyzer/values.yaml:228]: for sure, I was thinking it was more complex. thanks!

**Review by chipzoller — APPROVED** [review]

**Comment by chipzoller** [comment]
This should fully succeed after PR #2512 is merged.

File: cost-analyzer/charts/prometheus/templates/server-deployment.yaml [hunk 2]
```diff
@@ -59,10 +59,15 @@ spec:
...
           {{- end }}
           resources:
-{{ toYaml .Values.configmapReload.prometheus.resources | indent 12 }}
-          {{- with .Values.configmapReload.prometheus.containerSecurityContext }}
+          {{- toYaml .Values.configmapReload.prometheus.resources | nindent 12 }}
           securityContext:
-            {{- toYaml . | nindent 12 }}
+          {{- if .Values.global.containerSecurityContext }}
+            {{- toYaml .Values.global.containerSecurityContext | nindent 12 }}
+          {{- else if .Values.global.containerSecurityContext }}
```
Comment (thomasvn) [inline cost-analyzer/charts/prometheus/templates/server-deployment.yaml:66]: Note, the `if` and `else if ` statements here look to be the same

File: cost-analyzer/values.yaml [hunk 3]
```diff
@@ -217,8 +217,20 @@ global:
...
 
-  containerSecuritycontext: {}
-    # readOnlyRootFilesystem: true
+  securityContext:
+    runAsNonRoot: true
+    seccompProfile:
+      type: RuntimeDefault
+    fsGroup: 65534
+    runAsGroup: 65534
+    runAsUser: 65534
```
Comment (thomasvn) [inline cost-analyzer/values.yaml:226]: Any reason we've chosen `65534` here? Additionally, have we confirmed that all the containers in our chart can successfully run using a different user, with a `readonly` filesystem, and with all its capabilities dropped?

Comment (jessegoodier) [reply cost-analyzer/values.yaml:226]: yes, I did. 
the user works well for openshift and non-openshift tests I ran.

File: cost-analyzer/charts/prometheus/templates/server-deployment.yaml (hunk 2 above)
Comment (jessegoodier) [reply cost-analyzer/charts/prometheus/templates/server-deployment.yaml:66]: thanks for finding this!

File: cost-analyzer/values.yaml (hunk 3 above)
Comment (jessegoodier) [reply cost-analyzer/values.yaml:226]: also, I stole the idea from: https://github.com/prometheus-community/helm-charts/blob/main/charts/kube-prometheus-stack/values.yaml#L2286

Comment (thomasvn) [reply cost-analyzer/values.yaml:226]: Ah! According to stackexchange the significance of `65534` is that Linux reserves this for the `nobody` user/group. https://unix.stackexchange.com/questions/720557/what-is-the-id-of-nobody-user-and-nogroup-group

**Pull request merged**